### ISSUE TITLE
dealing with "I have no idea how this happened" multiple page/block bindings

### ISSUE THREAD

**Issue by Pomax** [cfpb/wagtail-inventory#43]
We were getting reports from our staff that some wagtail pages refused to publish and instead were throwing up server error 500s, and after investigating ([in this issue](https://github.com/mozilla/foundation.mozilla.org/issues/6162)) it turns out that somehow, due to whatever bizarre set of circumstances, some page/block type combinations have more than one row in the wagtailinventory table, so the `get_or_create` call in helper.py would end getting (rather than creating) more than one result, causing Django to crash because get_or_create expects zero or one results, not "zero or more" =)

So my (termporary) solution to this problem has been to change the helper.py code so that rather than a list comprehension:

```python
    return [
        PageBlock.objects.get_or_create(page=page, block=block)[0]
        for block in page_blocks
    ]
```

it does the more verbose "check if there are any results, then depending on the number of results, do one of three things":

```python
    list = []

    for block in page_blocks:
        bindings = PageBlock.objects.filter(page=page, block=block)

        if bindings.count() > 0:
            list.append(bindings.first())
            for index, entry in enumerate(bindings):
                if index > 0:
                    entry.delete()
        else:
            result = PageBlock.objects.create(page=page, block=block)
            list.append(result)

    return list
```

It's not as clean or efficient, but it does allow the code to do "what it should be doing" while also cleaning up any problem cases it runs into at the same time - while an unusual situation, it might be worth putting a solution like this in place in the official helper.py so that even if it runs into an unexpected database situation, it'll gracefully recover rather than trigger a server error 500

**Comment by chosak** [comment]
@Pomax thanks for this report! I can definitely imagine some ways this might happen; for example if two different users save a page at (close to) the same time. From a quick look it seems to me like some of the management of PageBlocks should probably take place inside transactions, to prevent this sort of thing from happening. Additionally, given the current data model there should maybe be a unique constraint on (page, block) -- this wouldn't solve the problem but would raise at error at save time which would at least indicate when something unexpected is happening.

**Comment by Pomax** [comment]
Happy to hear it's something worth improving!

That I have to say, I'm personally a fan of the "if it's broken and the code knows how to fix it, better to fix it and report that things got fixed than to throw an error", since at best an engineering team learns about it immediately and then figures out how to fix things in a few minutes, but more likely they'll hear about it after users have run into it a few times already, and then need a fair bit of time to figure out what's going on and how to fix that. At which point users won't have been able to work on potentially-on-a-deadline content for one or even two days =)

**Comment by chosak** [comment]
This should be fixed by the new migration in #45 that cleans up any multiple bindings and should properly prevent them in future.
